README.md
# Readable

Elixir language have standard `String.Chars` protocol which describes `T -> String` relation (like Haskell [Show](https://www.haskell.org/haddock/libraries/GHC.Show.html) type class). But Elixir don't have any standard protocol for `String -> T` relation (some equivalent of Haskell [Read](https://www.haskell.org/haddock/libraries/GHC.Read.html) type class). But I think such protocol completely make sense.

This protocol describes read relation between 2 types. Instance of type `B` can be read from instance of type `A`. Usually type `A = String`, but this implementation don't have this constraint.

<img src="priv/img/logo.png" width="300"/>

## Example

```elixir
import Read

defreadable Integer, from: x :: BitString do
  String.to_integer(x)
end

defreadable URI, from: x :: BitString do
  URI.parse(x)
end

defreadable NaiveDateTime, from: x :: Tuple do
  case NaiveDateTime.from_erl(x) do
    {:ok, y} -> y
    {:error, _} -> fail!(x)
  end
end
```

and then we can use implementation of `Read` type class

```elixir
iex> Read.read("123", Integer)
123

iex> Read.read("https://hello.world", URI)
%URI{
  authority: "hello.world",
  fragment: nil,
  host: "hello.world",
  path: nil,
  port: 443,
  query: nil,
  scheme: "https",
  userinfo: nil
}

iex> Read.read({{2000, 1, 1}, {13, 30, 15}}, NaiveDateTime)
~N[2000-01-01 13:30:15]

iex> Read.read({{2000, 13, 1}, {13, 30, 15}}, NaiveDateTime)
** (Readable.Exception) NaiveDateTime can not be read from {{2000, 13, 1}, {13, 30, 15}}

iex> Read.read("https://hello.world", TypeNotExist)
** (ArgumentError) argument error
    :erlang.binary_to_existing_atom("Elixir.Readable.From.BitString.To.TypeNotExist", :utf8)
    (elixir) src/elixir_aliases.erl:119: :elixir_aliases.safe_concat/1
    (readable) lib/read.ex:84: Read.read/2
```

Also there is strict macro version of helper which checks existence of target type in compile-time:

```elixir
iex> import Read
Read
iex> mk_read("123", Integer)
123
iex> mk_read("123", TypeNotExist)
** (UndefinedFunctionError) function TypeNotExist.__struct__/0 is undefined (module
 TypeNotExist is not available)
    TypeNotExist.__struct__()
    (typable 0.3.0) lib/typable.ex:94: Type.assert_exist!/1
    (readable 0.2.1) expanding macro: Read.mk_read/2
```

## Installation

The package can be installed by adding `readable` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:readable, "~> 0.3"}
  ]
end
```


CHANGELOG.md
# Changelog

## [v0.2.1](https://github.com/coingaming/readable/tree/v0.2.1) (2020-03-07)

[Full Changelog](https://github.com/coingaming/readable/compare/v0.2.0...v0.2.1)

## [v0.2.0](https://github.com/coingaming/readable/tree/v0.2.0) (2020-03-07)

[Full Changelog](https://github.com/coingaming/readable/compare/v0.1.0...v0.2.0)

## [v0.1.0](https://github.com/coingaming/readable/tree/v0.1.0) (2019-07-10)

[Full Changelog](https://github.com/coingaming/readable/compare/dfa29bed1d6ec62de5647d2a11fd1a6f89e9eaba...v0.1.0)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


test/readable_test.exs
defmodule ReadableTest do
  use ExUnit.Case
  doctest Read
  doctest Readable
end


mix.exs
defmodule Readable.MixProject do
  use Mix.Project

  def project do
    [
      app: :readable,
      version: "VERSION" |> File.read!() |> String.trim(),
      elixir: "~> 1.9",
      start_permanent: Mix.env() == :prod,
      consolidate_protocols: not (Mix.env() in [:dev, :test]),
      deps: deps(),
      aliases: aliases(),
      # excoveralls
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.travis": :test,
        "coveralls.circle": :test,
        "coveralls.semaphore": :test,
        "coveralls.post": :test,
        "coveralls.detail": :test,
        "coveralls.html": :test
      ],
      # dialyxir
      dialyzer: [
        ignore_warnings: ".dialyzer_ignore",
        plt_add_apps: [
          :mix,
          :ex_unit
        ]
      ],
      # ex_doc
      name: "Readable",
      source_url: "https://github.com/coingaming/readable",
      homepage_url: "https://github.com/coingaming/readable",
      docs: [main: "readme", extras: ["README.md"]],
      # hex.pm stuff
      description: "read type class families",
      package: [
        licenses: ["Apache 2.0"],
        files: ["lib", "priv", "mix.exs", "README*", "VERSION*"],
        maintainers: ["ILJA TKACHUK"],
        links: %{
          "GitHub" => "https://github.com/coingaming/readable",
          "Author's home page" => "http://itkach.uk"
        }
      ]
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:typable, "~> 0.3"},
      # development tools
      {:excoveralls, "~> 0.11", runtime: false, only: [:dev, :test]},
      {:dialyxir, "~> 0.5", runtime: false, only: [:dev, :test]},
      {:ex_doc, "~> 0.19", runtime: false, only: [:dev, :test]},
      {:credo, "~> 0.9", runtime: false, only: [:dev, :test]},
      {:boilex, "~> 0.2", runtime: false, only: [:dev, :test]}
    ]
  end

  defp aliases do
    [
      docs: ["docs", "cmd mkdir -p doc/priv/img/", "cmd cp -R priv/img/ doc/priv/img/", "docs"]
    ]
  end
end


lib/read.ex
defmodule Read do
  @moduledoc """
  Provides utilities to implement and work with `Readable` types
  """

  @doc """
  Helper to define readable relation for pair of types.
  Accepts:
  - target type which we want to read
  - source arg `expression :: type` pair
  - block of code where relation is described

  ## Examples

  ```
  iex> quote do
  ...>   import Read
  ...>   defreadable URI, from: x :: BitString do
  ...>     URI.parse(x)
  ...>   end
  ...> end
  ...> |> Code.compile_quoted
  iex> import Read
  iex> read("https://hello.world", URI)
  %URI{
    authority: "hello.world",
    fragment: nil,
    host: "hello.world",
    path: nil,
    port: 443,
    query: nil,
    scheme: "https",
    userinfo: nil
  }

  iex> quote do
  ...>   import Read
  ...>   defreadable NaiveDateTime, from: x :: Tuple do
  ...>     case NaiveDateTime.from_erl(x) do
  ...>       {:ok, y} -> y
  ...>       {:error, _} -> fail!(x)
  ...>     end
  ...>   end
  ...> end
  ...> |> Code.compile_quoted
  iex> import Read
  iex> read({{2000, 1, 1}, {13, 30, 15}}, NaiveDateTime)
  ~N[2000-01-01 13:30:15]
  iex> read({{2000, 13, 1}, {13, 30, 15}}, NaiveDateTime)
  ** (Readable.Exception) NaiveDateTime can not be read from {{2000, 13, 1}, {13, 30, 15}}
  ```
  """
  defmacro defreadable(
             quoted_to_type,
             [from: {:"::", _, [quoted_from_expression, quoted_from_type]}],
             do: code
           ) do
    {to_type, []} = Code.eval_quoted(quoted_to_type, [], __CALLER__)
    {from_type, []} = Code.eval_quoted(quoted_from_type, [], __CALLER__)
    :ok = Type.assert_exist!(to_type)
    :ok = Type.assert_exist!(from_type)

    type =
      [Readable, From, from_type, To, to_type]
      |> Module.concat()

    quote do
      defmodule unquote(type) do
        @enforce_keys [:data]
        defstruct @enforce_keys
      end

      defimpl Readable, for: unquote(type) do
        defp fail(x) do
          %Readable.Exception{
            message:
              unquote("#{inspect(to_type)} can not be read from ") <>
                inspect(x)
          }
        end

        defp fail!(x) do
          x
          |> fail()
          |> raise()
        end

        def read(%unquote(type){data: unquote(quoted_from_expression)}) do
          unquote(code)
        end
      end
    end
  end

  @doc """
  Helper to read expression into given type

  ## Examples

  ```
  iex> Read.read("1", Integer)
  1
  ```
  """
  def read(from_expression, to_type) do
    type =
      [
        Readable,
        From,
        Type.type_of(from_expression),
        To,
        to_type
      ]
      |> Module.safe_concat()

    %{__struct__: type, data: from_expression}
    |> Readable.read()
  end

  @doc """
  Strict macro helper to read expression into given type.
  Returns compile-time error if target type does not exist.

  ## Examples

  ```
  iex> import Read
  iex> mk_read("123", Integer)
  123
  iex> quote do
  ...>   import Read
  ...>   mk_read("123", TypeNotExist)
  ...> end
  ...> |> Code.eval_quoted()
  ** (UndefinedFunctionError) function TypeNotExist.__struct__/0 is undefined (module TypeNotExist is not available)
  ```
  """
  defmacro mk_read(from_expression, quoted_to_type) do
    {to_type, []} = Code.eval_quoted(quoted_to_type, [], __CALLER__)
    :ok = Type.assert_exist!(to_type)

    quote location: :keep do
      Read.read(unquote(from_expression), unquote(to_type))
    end
  end
end
